pkg/controller/nodeipam/ipam/multi_cidr_range_allocator.go

```diff
@@ -0,0 +1,1158 @@
+	}
+
+	if nodeList != nil {
+		for _, node := range nodeList.Items {
+			if len(node.Spec.PodCIDRs) == 0 {
+				klog.V(4).Infof("Node %v has no CIDR, ignoring", node.Name)
+				continue
+			}
+			klog.V(0).Infof("Node %v has CIDR %s, occupying it in CIDR map", node.Name, node.Spec.PodCIDRs)
+			if err := ra.occupyCIDRs(&node); err != nil {
+				// This will happen if:
+				// 1. We find garbage in the podCIDRs field. Retrying is useless.
+				// 2. CIDR out of range: This means ClusterCIDR is not yet created
+				// This error will keep crashing controller-manager until the
+				// appropriate ClusterCIDR has been created
+				return nil, err
+			}
+		}
+	}
+
+	nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: controllerutil.CreateAddNodeHandler(ra.AllocateOrOccupyCIDR),
+		UpdateFunc: controllerutil.CreateUpdateNodeHandler(func(_, newNode *v1.Node) error {
+			// If the PodCIDRs list is not empty we either:
+			// - already processed a Node that already had CIDRs after NC restarted
+			//   (cidr is marked as used),
+			// - already processed a Node successfully and allocated CIDRs for it
+			//   (cidr is marked as used),
+			// - already processed a Node but we saw a "timeout" response and
+			//   request eventually got through in this case we haven't released
+			//   the allocated CIDRs (cidr is still marked as used).
+			// There's a possible error here:
+			// - NC sees a new Node and assigns CIDRs X,Y.. to it,
+			// - Update Node call fails with a timeout,
+			// - Node is updated by some other component, NC sees an update and
+			//   assigns CIDRs A,B.. to the Node,
+			// - Both CIDR X,Y.. and CIDR A,B.. are marked as used in the local cache,
+			//   even though Node sees only CIDR A,B..
+			// The problem here is that in in-memory cache we see CIDR X,Y.. as marked,
+			// which prevents it from being assigned to any new node. The cluster
+			// state is correct.
+			// Restart of NC fixes the issue.
+			if len(newNode.Spec.PodCIDRs) == 0 {
+				return ra.AllocateOrOccupyCIDR(newNode)
+			}
+			return nil
+		}),
+		DeleteFunc: controllerutil.CreateDeleteNodeHandler(ra.ReleaseCIDR),
+	})
+
+	return ra, nil
+}
+
+func (r *multiCIDRRangeAllocator) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	klog.Infof("Starting Multi CIDR Range allocator")
+	defer klog.Infof("Shutting down Multi CIDR Range allocator")
+
+	if !cache.WaitForNamedCacheSync("multi_cidr_range_allocator", stopCh, r.nodesSynced, r.clusterCIDRSynced) {
+		return
+	}
+
+	// raWaitGroup is used to wait for the RangeAllocator to finish the goroutines.
+	var raWaitGroup sync.WaitGroup
+
+	for i := 0; i < cidrUpdateWorkers; i++ {
+		raWaitGroup.Add(1)
+		go func() {
+			defer raWaitGroup.Done()
+			r.worker(stopCh)
+		}()
+	}
+
+	raWaitGroup.Wait()
+
+	<-stopCh
+}
+
+func (r *multiCIDRRangeAllocator) worker(stopChan <-chan struct{}) {
+	for {
+		select {
+		case workItem, ok := <-r.nodeCIDRUpdateChannel:
+			r.lock.Lock()
```

Comment:
is this being used? I can't see anything adding workItems to the channel, seems that everything is processed from the EventHandlers ... EventHandlers block the infomer queue . it should not take more than 100ms to process an object on the handler see https://github.com/kubernetes/kubernetes/pull/103917

Reply:
wojtek comments is specially interesting https://github.com/kubernetes/kubernetes/pull/103917#discussion_r677157203